### PR TITLE
turn libkmod into a dlopen() dependency, too

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

**Comment by poettering** [comment]
@mrc0mmand any idea what i might be missing here? i updated test-functions with the libkmod dep, assuming that would make the testsuite work, but that wasn't enough?

**Comment by mrc0mmand** [comment]
> @mrc0mmand any idea what i might be missing here? i updated test-functions with the libkmod dep, assuming that would make the testsuite work, but that wasn't enough?

The Testing Farm error is unrelated (that's some post mass rebuild issue). For the Arch Linux one I'll need to prepare a fix for the mkinitcpio hook, since it needs to explicitly pull the libkmod dep in (https://gitlab.archlinux.org/archlinux/packaging/packages/systemd/-/blob/main/initcpio-install-systemd?ref_type=heads#L127). I'll restart the job once that's done.

**Comment by mrc0mmand** [comment]
Monkey-patched and restarted, hopefully it'll be content this time. Ignore the Testing Farm job for now, I'll try to deal with it later today.

**Comment by poettering** [comment]
@mrc0mmand thank you so much!

**Comment by mbiebl** [comment]
Using dlopen for more "exotic"  / less frequently used dependencies seems ok. It does come at a cost though. As you no longer get any automatic shlibs dependency tracking.

In case of libkmod I'm not actually sure if the dlopen approach is the desirable one.

**Comment by poettering** [comment]
> Using dlopen for more "exotic" / less frequently used dependencies seems ok. It does come at a cost though. As you no longer get any automatic shlibs dependency tracking.

as mentioned elsewhere before: i am happy to extend ELF metadata declaring such weak deps if there's interest. As i understand people weren't intereste din that (did that change?), but it's the bed you made yourself then.

There's nothing stopping anyone to declare weak deps in ELF notes or so, except politics i guess..

> In case of libkmod I'm not actually sure if the dlopen approach is the desirable one.

libkmod is of no use in a container, and it has a surprisingly large dep tree, see the lddtree output in the commit msg. It actually makes a difference...

**Comment by mbiebl** [comment]
none of the deps pulled in by libkmod (zstd,lzma) aren't already pulled in by journald.
If I'm missing something, which part of the "large dependency tree" exactly are you referring to?

**Comment by mbiebl** [comment]
> > Using dlopen for more "exotic" / less frequently used dependencies seems ok. It does come at a cost though. As you no longer get any automatic shlibs dependency tracking.
> 
> as mentioned elsewhere before: i am happy to extend ELF metadata declaring such weak deps if there's interest. As i understand people weren't intereste din that (did that change?), but it's the bed you made yourself then.

It seems you failed to convince maintainers of the relevant domains. This is not something I can comment on, as I was not involved.

**Comment by DaanDeMeyer** [comment]
@poettering Obviously you have to do the dlopen() dance for compression deps as well to make this worth it (which would happen to also be very useful for mkosi-initrd ;) )

**Comment by mbiebl** [comment]
Fwiw, I'd be happy to see more elaborate support for weak ELF deps, as it could be really useful.
But it doesn't exist as of today.

**Comment by mrc0mmand** [comment]
I have a feeling that this breaks Ubuntu CI as well, in a similar fashion to the Arch Linux job:

```
3188s [    2.468538] Btrfs loaded, crc32c=crc32c-vpmsum, zoned=yes, fsverity=yes
3188s Scanning for Btrfs filesystems
3188s done.
3188s Begin: Waiting for root file system ... Begin: Running /scripts/local-block ... done.
3188s done.
3188s Gave up waiting for root file system device.  Common problems:
3188s  - Boot args (cat /proc/cmdline)
3188s    - Check rootdelay= (did the system wait long enough?)
3188s  - Missing modules (cat /proc/modules; ls /dev)
3188s ALERT!  UUID=0a02b638-bd49-4573-80f4-e676eaf07236 does not exist.  Dropping to a shell!
```

/cc @bluca

**Comment by poettering** [comment]
> @poettering Obviously you have to do the dlopen() dance for compression deps as well to make this worth it (which would happen to also be very useful for mkosi-initrd ;) )

It's already on the TODO list. And yes, I want to in particular the more exotic ones to be dlopen() deps

**Comment by poettering** [comment]
> It seems you failed to convince maintainers of the relevant domains. This is not something I can comment on, as I was not involved.

Well, I am not sure it's possible to convince the dpkg people of anything, is it? ;-)

**Comment by poettering** [comment]
> Fwiw, I'd be happy to see more elaborate support for weak ELF deps, as it could be really useful. But it doesn't exist as of today.

Well, but I am sure that shouldn't hold us back in systemd to make the dependency chain smaller. And yes, I think there's some more deps we should cover like this. Specifically the compression libs, i.e. libbz2, libz4 and suchlike at least. And also libpam. So expect more like this sooner or later.

I'd be happy to make systemd expose these weak deps automatically in some metadata, that's all I am saying, but I am not going to champion this with the packaging tool maintainers, I got enough on my plate. All I am saying is that if there's interest to have this all handled automatically, I'd be happy to do my part on the systemd side. But on the packaging tools side someone else has to do the stuff.

It probably could be as easy as having some macros in systemd that insert a new ELF section with this data, and then have some readelf script that extracts this at package build time.

**Comment by bluca** [comment]
> > It seems you failed to convince maintainers of the relevant domains. This is not something I can comment on, as I was not involved.
> 
> Well, I am not sure it's possible to convince the dpkg people of anything, is it? ;-)

Yes, but we don't need to wait for them fortunately, we can just do whatever we want in our packages

**Comment by bluca** [comment]
> > Fwiw, I'd be happy to see more elaborate support for weak ELF deps, as it could be really useful. But it doesn't exist as of today.
> 
> Well, but I am sure that shouldn't hold us back in systemd to make the dependency chain smaller. And yes, I think there's some more deps we should cover like this. Specifically the compression libs, i.e. libbz2, libz4 and suchlike at least. And also libpam. So expect more like this sooner or later.
> 
> I'd be happy to make systemd expose these weak deps automatically in some metadata, that's all I am saying, but I am not going to champion this with the packaging tool maintainers, I got enough on my plate. All I am saying is that if there's interest to have this all handled automatically, I'd be happy to do my part on the systemd side. But on the packaging tools side someone else has to do the stuff.
> 
> It probably could be as easy as having some macros in systemd that insert a new ELF section with this data, and then have some readelf script that extracts this at package build time.

Going via the ELF route is a good idea, however we don't need new sections and new tools - we have the packaging metadata spec already implemented, with all the linkers supporting inserting the section correctly, and we can just use that, adding a new field. It's json and it's extendable exactly for this purpose.

**Comment by poettering** [comment]
> Going via the ELF route is a good idea, however we don't need new sections and new tools - we have the packaging metadata spec already implemented, with all the linkers supporting inserting the section correctly, and we can just use that, adding a new field. It's json and it's extendable exactly for this purpose.

But that's added at package build-time, no? This stuff here is supposed to be added in the sources, and *consumed* at package build-time. I don't see how that overlaps?

**Comment by bluca** [comment]
> > Going via the ELF route is a good idea, however we don't need new sections and new tools - we have the packaging metadata spec already implemented, with all the linkers supporting inserting the section correctly, and we can just use that, adding a new field. It's json and it's extendable exactly for this purpose.
> 
> But that's added at package build-time, no? This stuff here is supposed to be added in the sources, and _consumed_ at package build-time. I don't see how that overlaps?

Any ELF section is added at build time, pretty much by definition, no?

**Comment by poettering** [comment]
> Any ELF section is added at build time, pretty much by definition, no?

Well. Not really. if we denote the ELF section to be generated via `__attribute__((__section__(x)))` in our sources, then I wouldn't really call this "at build time", but "at development time". 

The packaging metadata spec is inserted at distro build time and carries distro info.

This is different from what we need for the dlopen() stuff: the developer writes it at development time, and it carries generic info that applies to all distros.

I really wouldn't try to merge this, as it fucks of the provenience rules. 

Make it clean: 

* `.note.package` ← added by distro builders with only distro info
* the new section ← added by upstream developers, with only upstream info

**Comment by bluca** [comment]
It cannot be in the sources, because it cannot be added by the developer, it needs to be added at build time. The information required depends on the build environment to be useful - ie, which package is being used. In the sources you can say "this uses kmod" but that's pretty much useless and provides no benefit over the status quo, as we already know it's used.

**Comment by mrc0mmand** [comment]
Regarding the test fails, I think there's a bug in udev somewhere, because it seems to segfault when libkmod is missing. I feel that it should just fail, not crash:

```
[    6.909429] systemd-modules-load[211]: Failed to initialize libkmod context: Operation not supported
[    8.990402] systemd-udevd[355]: Failed to initialize libkmod context: Operation not supported
[   14.560523] systemd-cryptsetup[424]: Key file /run/systemd/cryptsetup/keydev-test24_varcrypt/keyfile is world-readable. This is not a good idea!
[   14.675314] systemd-cryptsetup[424]: /run/systemd/cryptsetup/keydev-test24_varcrypt/keyfile has 0644 mode that is too permissive, please adjust the ownership and access mode.
[   17.673832] systemd-coredump[489]: Failed to connect to coredump service: No such file or directory
[   17.701781] systemd[1]: systemd-udevd.service: Main process exited, code=dumped, status=11/SEGV
[   17.703629] systemd[1]: systemd-udevd.service: Failed with result 'core-dump'.
```

**Comment by mrc0mmand** [comment]
I managed to get _some_ stack trace from the crash:

```gdb
(gdb) bt full
#0  0x0000000000000000 in ?? ()
No symbol table info available.
#1  0x000055c2991ddf3f in builtin_kmod_exit () at ../src/udev/udev-builtin-kmod.c:66
        __func__ = "builtin_kmod_exit"
#2  0x000055c2991e28a3 in udev_builtin_exit () at ../src/udev/udev-builtin.c:52
        i = <optimized out>
#3  udev_builtin_exit () at ../src/udev/udev-builtin.c:46
        i = <optimized out>
#4  0x000055c2991cd903 in manager_free (manager=0x55c299ea9960) at ../src/udev/udev-manager.c:141
No locals.
#5  0x000055c2991b5a5f in manager_free (manager=0x55c299ea9960) at ../src/udev/udev-manager.c:138
No locals.
#6  manager_freep (p=<optimized out>) at ../src/udev/udev-manager.h:57
No locals.
#7  run_udevd (argv=<optimized out>, argc=<optimized out>) at ../src/udev/udevd.c:289
        manager = 0x55c299ea9960
        fd_ctrl = <optimized out>
        fd_uevent = <optimized out>
        r = <optimized out>
        manager = <optimized out>
        fd_ctrl = <optimized out>
        fd_uevent = <optimized out>
        r = <optimized out>
        __func__ = <optimized out>
        _error = <optimized out>
        _level = <optimized out>
        _e = <optimized out>
        _error = <optimized out>
        _level = <optimized out>
        _e = <optimized out>
        _error = <optimized out>
        _level = <optimized out>
        _e = <optimized out>
        _error = <optimized out>
        _level = <optimized out>
        _e = <optimized out>
        pid = <optimized out>
        _level = <optimized out>
        _e = <optimized out>
        _error = <optimized out>
        _level = <optimized out>
        _e = <optimized out>
        _error = <optimized out>
        _level = <optimized out>
        _e = <optimized out>
#8  run (argv=<optimized out>, argc=<optimized out>) at ../src/udev/udevadm.c:124
        r = <optimized out>
        r = <optimized out>
#9  main (argc=<optimized out>, argv=<optimized out>) at ../src/udev/udevadm.c:140
        r = <optimized out>
        __func__ = "main"
```

**Comment by poettering** [comment]
> It cannot be in the sources, because it cannot be added by the developer, it needs to be added at build time. The information required depends on the build environment to be useful - ie, which package is being used. In the sources you can say "this uses kmod" but that's pretty much useless and provides no benefit over the status quo, as we already know it's used.

Well, we'd note the library name as passed to dlopen() there (i.e. "linkmod.so.2"). It would then be directly converted to a "Requires: libkmod.so.2()(64bit)" deps or so in RPM.

What would you expect we output instead? 

I mean, we might are apparently not on the same page at all here. if it's not the data that is encoded in the ldopen(), then what data specifically do you have in mind at all? And if it's added by the packager, then why export anything in systemd at all? i don't get it?

**Comment by mrc0mmand** [comment]
I patched the Rawhide job after I made sure udev no longer crashes without libkmod, so it should, hopefully, pass this time. Nevertheless, with this PR both the systemd dracut module (cc @keszybz) and the initcpio systemd hook (@eworm-de) will need to be updated to install libkmod explicitly once this change makes it to the respective distributions. And I suppose something similar will be needed for initramfs-tools as well for Ubuntu/Debian.

**Comment by eworm-de** [comment]
https://gitlab.archlinux.org/archlinux/packaging/packages/systemd/-/commit/6671472ddc84679520438b97eddea075bdf4dcc2

**Comment by poettering** [comment]
> It cannot be in the sources, because it cannot be added by the developer, it needs to be added at build time. 

Hmm? It *is* already in the sources, and added by the developer:

```$ git grep -A 10  dlopen | grep \\.so
src/locale/xkbcommon-util.c-                        &xkbcommon_dl, "libxkbcommon.so.0", LOG_DEBUG,
src/shared/bpf-dlopen.c:        dl = dlopen("libbpf.so.1", RTLD_LAZY);
src/shared/bpf-dlopen.c:                dl = dlopen("libbpf.so.0", RTLD_LAZY);
src/shared/bpf-dlopen.c-                                               "neither libbpf.so.1 nor libbpf.so.0 are installed: %s", dlerror());
src/shared/cryptsetup-util.c-                        &cryptsetup_dl, "libcryptsetup.so.12", LOG_DEBUG,
src/shared/elf-util.c-                        &dw_dl, "libdw.so.1", LOG_DEBUG,
src/shared/elf-util.c-                        &elf_dl, "libelf.so.1", LOG_DEBUG,
src/shared/firewall-util-iptables.c-                        "libip4tc.so.2", LOG_DEBUG,
src/shared/idn-util.c-                        &idn_dl, "libidn2.so.0", LOG_DEBUG,
src/shared/idn-util.c:        dl = dlopen("libidn.so.12", RTLD_LAZY);
src/shared/idn-util.c:                dl = dlopen("libidn.so.11", RTLD_LAZY);
src/shared/libarchive-util.c-                        "libarchive.so.13",
src/shared/libfido2-util.c-                        &libfido2_dl, "libfido2.so.1", LOG_DEBUG,
src/shared/module-util.c-                        "libkmod.so.2",
src/shared/password-quality-util-passwdqc.c-                        &passwdqc_dl, "libpasswdqc.so.1", LOG_DEBUG,
src/shared/password-quality-util-pwquality.c-                        &pwquality_dl, "libpwquality.so.1", LOG_DEBUG,
src/shared/pcre2-util.c-                        &pcre2_dl, "libpcre2-8.so.0", LOG_ERR,
src/shared/pkcs11-util.c-                        "libp11-kit.so.0", LOG_DEBUG,
src/shared/qrcode-util.c-        FOREACH_STRING(s, "libqrencode.so.4", "libqrencode.so.3") {
src/shared/tpm2-util.c-                        &libtss2_esys_dl, "libtss2-esys.so.0", LOG_DEBUG,
src/shared/tpm2-util.c-                        &libtss2_rc_dl, "libtss2-rc.so.0", LOG_DEBUG,
src/shared/tpm2-util.c-                        &libtss2_mu_dl, "libtss2-mu.so.0", LOG_DEBUG,
src/shared/userdb.c:        dl = dlopen(LIBDIR "/libnss_systemd.so.2", RTLD_LAZY|RTLD_NODELETE);
src/shared/userdb.c:                log_debug("Failed to dlopen(libnss_systemd.so.2), ignoring: %s", dlerror());
```

All I am suggesting is that the names of the `.so` fields listed here are also placed in some ELF section automatically via some macro magic.

> The information required depends on the build environment to be useful - ie, which package is being used. In the sources you can say "this uses kmod" but that's pretty much useless and provides no benefit over the status quo, as we already know it's used.

I am not following at all. I mean, it would be great if pkgconfig would export the actual soname a library is found under. Alas, it doesn't.

**Comment by keszybz** [comment]
> Nevertheless, with this PR both the systemd dracut module (cc @keszybz) and the initcpio systemd hook (@eworm-de) will need to be updated to install libkmod explicitly, once this change makes it to the respective distributions. And I suppose something similar will be needed for initramfs-tools as well for Ubuntu/Debian.

I'm not the right person for dracut, I have not to do with.

**Comment by DaanDeMeyer** [comment]
Guess which initramfs generation tool needs zero changes to accomodate this change? ;)

**Comment by keszybz** [comment]
> Guess which initramfs generation tool needs zero changes to accomodate this change? ;)

That's cheating. We need to update the packaging of systemd.

**Comment by keszybz** [comment]
→ https://src.fedoraproject.org/rpms/systemd/pull-request/125

**Comment by keszybz** [comment]
Fedora rawhide fails because mkosi hasn't been released with the patch for gpg key handling.
But the jammy-* seems to fail when trying to load modules.
So I guess we can't merge this until the CI is reworked somehow.

**Comment by bluca** [comment]
So as the tests show, there's no way this can be anything but a hard dependency in the package, given systems won't boot without the library, so it can't really be skipped. Is it really worth it to make it a dlopen?

**Comment by DaanDeMeyer** [comment]
@bluca Containers don't need kmod

**Comment by DaanDeMeyer** [comment]
@bluca Containers don't need kmod, so you make it a recommends for systemd and a requires for udev (see https://src.fedoraproject.org/rpms/systemd/pull-request/125).

**Comment by bluca** [comment]
I know they don't, but the package needs to have a dependency, as recommends can and are often disabled even when installing full systems, so realistically you can't really build a containers from packages without it - only if you are building from sources specifically for a container installation, but then it can just be disabled in meson?

**Comment by DaanDeMeyer** [comment]
> I know they don't, but the package needs to have a dependency, as recommends can and are often disabled even when installing full systems, so realistically you can't really build a containers from packages without it - only if you are building from sources specifically for a container installation, but then it can just be disabled in meson?

See my updated answer, the hard dependency needs to be added on udev, not systemd. A system won't boot without udev either, so it's perfectly fine to move the kmod hard depend from systemd to udev

**Comment by bluca** [comment]
Ok that might work, let's see what breaks

**Comment by YHNdnzj** [comment]
> > @poettering Obviously you have to do the dlopen() dance for compression deps as well to make this worth it (which would happen to also be very useful for mkosi-initrd ;) )
> 
> It's already on the TODO list. And yes, I want to in particular the more exotic ones to be dlopen() deps

Hmm, I wonder if we can just switch everything to libarchive? It was added as a dlopen()-ed dep recently (https://github.com/systemd/systemd/pull/31075), while journald is still linked to zstd/xz/...

**Comment by YHNdnzj** [comment]
If libarchive is sufficient for our compression work, that is. It should be easier to eliminate direct deps on each compression lib we make use of this way.

**Comment by poettering** [comment]
> > @poettering Obviously you have to do the dlopen() dance for compression deps as well to make this worth it (which would happen to also be very useful for mkosi-initrd ;) )
> 
> It's already on the TODO list. And yes, I want to in particular the more exotic ones to be dlopen() deps

→ #31550

File: src/shared/module-util.h
Comment (daandemeyer) [inline src/shared/module-util.h:54]: I think we do this inline in the function itself these days instead of doing multiple prototypes? See other dlopen libs implementation.

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

File: src/shared/module-util.h
Comment (yuwata) [reply src/shared/module-util.h:54]: I prefer the current style, and I'd like to convert other EOPNOTSUPP cases into inline. But, not a strong opinion. I think it is OK to keep this at least in this PR.

**Review by yuwata — APPROVED** [review]
LGTM. But, needs rebase again.

File: src/shared/module-util.h
Comment (poettering) [reply src/shared/module-util.h:54]: hmm, this is actually a mixed back. pkcs11, tpm2 all do the static inline stuff.

i have no idea what style is the better approach. With LTO the two ways should have similar effects. 

i can see pros/cons on both approaches:

1. static inline approach means fewer ifdef
2. static inline approach means ifdef in header though, which means they aren't hidden away
3. static inline approach means it should be optimized away even without LTO

hence, not clear cut to me.

**Comment by poettering** [comment]
Force pushed a rebased version. Left the static inline stuff as is, since it's not clear at all to me which approach is the better one.

**Comment by poettering** [comment]
i upgaded the green label. if we agree that one of the two styles is better we can switch everything over, but I am sure that should happen in a separate PR.

But first we have to figure out which style actually is better. I really have no idea.

File: src/shared/module-util.c
Comment (YHNdnzj) [inline src/shared/module-util.c]: Hmm, if `_printf_` is used this shouldn't be necessary?

File: src/shared/module-util.c
Comment (YHNdnzj) [inline src/shared/module-util.c]: This is used only once, so maybe just inline the flag while at it?

File: src/shared/module-util.c
Comment (YHNdnzj) [inline src/shared/module-util.c]: spurious newline?

File: src/shared/module-util.c
Comment (YHNdnzj) [inline src/shared/module-util.c:188]: assert(ret);

File: src/udev/udev-builtin-kmod.c
Comment (YHNdnzj) [inline src/udev/udev-builtin-kmod.c:60]: This should be moved down?

File: src/udev/udev-builtin-kmod.c
Comment (YHNdnzj) [inline src/udev/udev-builtin-kmod.c:64]: Hmm, other builtins don't seem to log above debug in `_init`?

**Comment by bluca** [comment]
This breaks booting in jammy-amd64 - and I have no idea why

**Comment by poettering** [comment]
Educated guess: kmod is no longer pulled into the initrd and ubuntu compiles some kernel functionality we need as kmod which cannot be no longer loaded now.

**Comment by bluca** [comment]
There's no systemd in the initrd though, and udev has a manual dependency on libkmod2

**Comment by poettering** [comment]
So, the opensuse Ci test definitely fails because kmod is not available in the mkosi image.

**Comment by bluca** [comment]
I'll try to repro the Ubuntu one

File: src/udev/udev-builtin-kmod.c
Comment (poettering) [reply src/udev/udev-builtin-kmod.c:60]: hmm, where to?

File: src/udev/udev-builtin-kmod.c
Comment (poettering) [reply src/udev/udev-builtin-kmod.c:64]: no quite true, the netif rename on calls naming_scheme() which logs at various higher levels.

File: src/udev/udev-builtin-kmod.c
Comment (YHNdnzj) [reply src/udev/udev-builtin-kmod.c:60]: Previously this is logged only if `kmod_new` succeeds, so this should stay below `module_setup_context`?

File: src/udev/udev-builtin-kmod.c
Comment (poettering) [reply src/udev/udev-builtin-kmod.c:64]: and the code around the udev_builtin_init() invocations tends to log loudly too.

i'd thus argue, we actually should fix the others to log above LOG_DEBUG too

File: src/udev/udev-builtin-kmod.c
Comment (poettering) [reply src/udev/udev-builtin-kmod.c:60]: sure, but that's a minor detail. the kmod_new() is a cheap call, that just malloc()s a bit, and never really fails. the heavy call is kmod_load_resources(), which loads stuff of disk. I think the debug message should tell us that this heavy operation is now initiating, hence the log message should be *before* it. The fact that it is now also before kmod_new() doesn't really matter much I'd say.

**Comment by poettering** [comment]
Force pushed a new version btw, hopefully passes on suse.

**Comment by DaanDeMeyer** [comment]
@fbuihuu Can you add an explicit dependency on kmod to the udev package in the opensuse systemd spec (maybe gated behind the %upstream macro to accomodate for this change?

File: src/udev/udev-builtin-kmod.c
Comment (YHNdnzj) [reply src/udev/udev-builtin-kmod.c:60]: OK. That makes sense.

**Comment by bluca** [comment]
boot is still broken on suse, stuck waiting for devices

**Comment by poettering** [comment]
maybe adding likmod2 instead of kmod as rpm to opensuse helps? trying again

**Comment by DaanDeMeyer** [comment]
@poettering Note that you can `dnf install zypper` and then do `mkosi -d opensuse -f qemu` to reproduce locally.

**Comment by bluca** [comment]
jammy-amd64 is back to working in other PRs, trying again here

**Comment by poettering** [comment]
So I take it ubuntu jammy is using their own initrd generator which does not run systemd, but does run systemd-udevd?

I guess it needs to be fixed to pull in libkmod then probably. How can we do this?

**Comment by bluca** [comment]
Ah of course it's initramfstools and it does ldd based resolution like dracut goddamnit

**Comment by keszybz** [comment]
> Ah of course it's initramfstools and it does ldd based resolution like dracut goddamnit

So we're waiting for an update there I guess.

**Review by keszybz — APPROVED** [review]
LGTM, but we need to get the packaging in order.

**Comment by keszybz** [comment]
`mkosi / ci (ubuntu, noble)`:
```
The following packages have unmet dependencies:
 libtss2-esys-3.0.2-0t64 : Breaks: libtss2-esys-3.0.2-0 (< 4.0.1-7.1ubuntu5)
 libtss2-mu-4.0.1-0t64 : Breaks: libtss2-mu-4.0.1-0 (< 4.0.1-7.1ubuntu5)
 libtss2-rc0t64 : Breaks: libtss2-rc0 (< 4.0.1-7.1ubuntu5)
 libtss2-tcti-device0t64 : Breaks: libtss2-tcti-device0 (< 4.0.1-7.1ubuntu5)
E: Unable to correct problems, you have held broken packages.
```
Hmmmmmmmmmmmm.

**Comment by DaanDeMeyer** [comment]
> `mkosi / ci (ubuntu, noble)`:
> 
> ```
> The following packages have unmet dependencies:
>  libtss2-esys-3.0.2-0t64 : Breaks: libtss2-esys-3.0.2-0 (< 4.0.1-7.1ubuntu5)
>  libtss2-mu-4.0.1-0t64 : Breaks: libtss2-mu-4.0.1-0 (< 4.0.1-7.1ubuntu5)
>  libtss2-rc0t64 : Breaks: libtss2-rc0 (< 4.0.1-7.1ubuntu5)
>  libtss2-tcti-device0t64 : Breaks: libtss2-tcti-device0 (< 4.0.1-7.1ubuntu5)
> E: Unable to correct problems, you have held broken packages.
> ```
> 
> Hmmmmmmmmmmmm.

Not our problem

**Comment by bluca** [comment]
Noble is still borked and is starting to stabilize. But yes it would be broken because of initramfstools, like jammy is showing. A temporary quick solution would be to link udev to libkmod, as you don't install udev in containers anyway normally

**Comment by keszybz** [comment]
> Noble is still borked and is starting to stabilize. But yes it would be broken because of initramfstools, like jammy is showing. A temporary quick solution would be to link udev to libkmod, as you don't install udev in containers anyway normally

That's not so easy. We have `--as-needed` probably, which means that if we add the linkage without any use, it'd get dropped anyway.

**Comment by poettering** [comment]
I am kinda tempted to just merge the thing, and ignore the CI for a bit, until somebody fixes that initramfs builder?

**Comment by bluca** [comment]
Please, no, that will make it impossible to get the CI changes done for your other PR, among other things. I'll work out a solution tonight or tomorrow.

**Comment by bluca** [comment]
I have pushed a fix, let's see if it works

**Comment by bluca** [comment]
It works, fortunately the udev hook is in the systemd package repo, so the change is self-contained and doesn't have to come from the initramfs package itself. Jammy is now working, noble has unrelated issues that have nothing to do with this.

**Review by bluca — APPROVED** [review]

**Pull request merged**
